src/lsm/cache_map.zig

```diff
@@ -203,73 +194,96 @@ pub fn CacheMapType(
+                    } else {
+                        // Case 3b.
+                        self.scope_rollback_log.appendAssumeCapacity(
+                            tombstone_from_key(key),
+                        );
+                    }
+                }
+            } else {
+                const gop = self.stash.getOrPutAssumeCapacity(value.*);
+                if (gop.found_existing) {
                     // Case 3a.
-                    self.scope_rollback_log.appendAssumeCapacity(stash_value);
+                    self.scope_rollback_log.appendAssumeCapacity(gop.key_ptr.*);
                 } else {
                     // Case 3b.
                     self.scope_rollback_log.appendAssumeCapacity(
-                        tombstone_from_key(key_from_value(value)),
+                        tombstone_from_key(key),
                     );
                 }
+                gop.key_ptr.* = value.*;
             }
         }
 
-        pub fn remove(self: *Self, key: Key) void {
+        pub fn remove(self: *CacheMap, key: Key) void {
             // The only thing that tests this in any depth is the cache_map fuzz itself.
             // Make sure we aren't being called in regular code without another once over.
             assert(constants.verify);
 
-            const maybe_removed = self.cache.remove(key);
+            if (self.cache) |*cache| {
+                const maybe_removed = cache.remove(key);
 
-            if (self.scope_is_active) {
-                // TODO: Actually, does the fuzz catch this...
-                // TODO: So if we delete from stash_2 and put to stash_1, there's a problem;
-                //       because when we undo our scope we insert back to stash_1 :/
-                self.scope_rollback_log.appendAssumeCapacity(
-                    maybe_removed orelse
-                        self.stash_1.getKey(tombstone_from_key(key)) orelse
-                        self.stash_2.getKey(tombstone_from_key(key)) orelse return,
-                );
-            }
+                if (self.scope_is_active) {
+                    // TODO: Actually, does the fuzz catch this...
+                    self.scope_rollback_log.appendAssumeCapacity(
+                        maybe_removed orelse
+                            self.stash.getKey(tombstone_from_key(key)) orelse return,
+                    );
+                }
 
-            // We always need to try remove from the stash; since it could have a stale value.
-            // The early return above is OK - if it doesn't exist, there's nothing to remove.
-            _ = self.stash_1.remove(tombstone_from_key(key));
-            _ = self.stash_2.remove(tombstone_from_key(key));
+                // We always need to try remove from the stash; since it could have a stale value.
+                // The early return above is OK - if it doesn't exist, there's nothing to remove.
+                _ = self.stash.remove(tombstone_from_key(key));
```

Comment:
Drive by, but I think we are doing extra hash map lookup here!

```
self.stash.getKey(tombstone_from_key(key))
self.stash.remove(tombstone_from_key(key));
```

these two will needlessly do the same work twice. If we are removig from stash unconditonally, mabybe pull `fetchRemove` to the top once, like we do in the other branch?